### DISCUSSION TITLE
Announcing Schema Validation (private beta)

### DISCUSSION THREAD

**Discussion by nvie** [liveblocks/liveblocks#674] [Announcements]
We are excited to announce our newest feature: [Schema Validation](https://liveblocks.io/docs/guides/schema-validation) (in private beta)! By defining a schema for your data, we’ll automatically validate all storage mutations in your Liveblocks rooms against your schema, ensuring data consistency, and reducing the risk of errors.

You can now sign up for our [private beta waitlist](https://docs.google.com/forms/d/197IrpiyeqlJo9xOCOfzwdcwBIk6nisk9pToJe3st3Ng/viewform). As a beta user, you will have the opportunity to test drive the feature and provide valuable feedback to help shape its development.

Also, feel free to leave a comment below if you have any questions or feedback for us—we’re listening! 🙏

**Comment by longergrass** [comment]
![Screenshot](https://github.com/liveblocks/liveblocks/assets/32229274/d5f507de-f013-4674-b92c-38f7122e3d5a)

I see on the home page it says Schema Validation is only available for Enterprise users (also webhooks).

Does this mean that once Schema Validation comes out of beta it will only be available for enterprise users? If so, what would the pricing look like for the enterprise plan?

**Reply by stevenfabre** [reply]
Hi there, webhooks is going out beta shortly and will be available on all paid plans. Schema validation is currently only available on the Enterprise plan. The pricing for that varies depending on your requirements and usage. Happy to walk you through it and hear more about your use case on a call.

When would be a good time for you to meet? Feel free to email us at hello@liveblocks.io and we can schedule something.

**Comment by Raid55** [comment]
I will add my requests and desired workflows here, I know I already shared these over a call but I am adding them here for posterity:

**CI/CD Integration**
The best way to keep track of schemas would be with a git repo, and ideally this is all automated through our github actions pipeline. I am never a fan of using a 3rd party UI to do something that should be processed automatically. The idea workflow would allow us to define liveblocks schema structure that would also allow for schemas to be assigned automatically to rooms with "certain metadata".

This CI/CD tool will also need to take the environment the schema is being built for (see more below).

The real kicker here is then a schema for room "metadata" to guarantee that all of our rooms have the appropriate metadata on creation instead of relying on room names with delimiters. We will always want every room we have to have the workspaceID, the room type, and if the room is archived or not (soft delete).

Lastly, it would be nice if the CI/CD tool (it could be a github app tbh, but i'm biassed cuz we use gh actions) would also have a verification phase so that when a schema is about to be updated the pipeline can output a message in our slack and ping the correct engineer for verification. We generally always try to talk about our schema changes before allowing them to be merged in. Ofc this is not super important since this is solvable with a short script.

**Schema Migrations**
The Ability to write code that will migrate one schema to another.
When a room in V1 gets migrated to V2, it can still accept updates from a v1 schema but will auto migrate them to v2.
Migrating schemas should be able to target room metadata and perform batch room migrations. Room Metadata seems more reliable than room ID's with delimiters.

To go further into that last point around metadata. We would really like to automate as much as possible, new rooms with specific metadata ("type": "Board") would automatically get assigned the Board schema at the latest version avail.

**Rollbacks**
With Schema Migrations also comes a robust rollback system. Ideally it would be tied to the tool being used and when we initiate an app rollback it would also rollback liveblocks schema to the previous version.  A rollback may not always be necessary:
App V1.0.4 - Schema 01
App V1.0.5 - Schema 01
App v1.1.2 - Schema 02 

**Schema Env**
Testing out new features in our staging env. should not affect schemas that are on production.

**Discriminating Unions**
We have a few data structures that take advantage of this and it really makes huge developer experience diff.

**YupJS or Alt Validation**
some custom validators would be nice, Yup allows for custom validators to be written and also use a bunch of built-in ones that would make data integrity more integriful. (lol)

**Write Once**
This is not a point since it seems like its already the way you are implementing this but I really dont want to keep track of two schemas, if we write it once, its reusable both on our backend and front end (for typescript).

Right now it feels like the schema system is not powerful enough to support the processes we want so I am being apprehensive about adding them. However i'm sure its only going to improve and eventually become an indispensable part of working with Liveblocks.

**Comment by kyleschaeffer** [comment]
Schema validation is great, thank you! We would most definitely appreciate a schema migration feature. We've experimented with migration scripts, but with so many rooms in production, it would take several hours to complete a single migration, making it a rather tenuous solution. Anything that we could do in a shorter amount of time would be extremely valuable.
